numpy/core/src/multiarray/mapping.c

```diff
@@ -285,6 +304,12 @@ prepare_index(PyArrayObject *self, PyObject *index,
     int i;
     npy_intp n;
 
+    PyObject *index = maybe_apply_callable(self, arg);
+    Py_DECREF(arg);
```

Comment:
This refcounting seems shady to me. `arg` is passed in as a borrowed reference, what we need to do now is rather DECRF `index` when done with it, which is probably directly after `unpack_indices`.